File: lint-arwaky/src-python/capabilities/arch_internal_checker.py
"""arch_internal_checker — Internal architectural rule checks (barrels, primitives)."""

from ..taxonomy import (
    ColumnNumber,
    ErrorCode,
    FilePath,
    FilePathList,
    LintMessage,
    LintResult,
    LintResultList,
    LineNumber,
    PRIMITIVE_TYPE_LIST,
    PrimitiveTypeList,
    Severity,
    AdapterName,
    LayerNameVO,
)
from ..contract import IInternalCheckerProtocol


class ArchInternalChecker(IInternalCheckerProtocol):
    """Handles internal layer rules like barrel completeness and primitive usage."""

    def check_layer_internal_rules(
        self,
        analyzer,
        files: FilePathList,
        root_dir: FilePath,
        results: LintResultList,
    ) -> None:
        for f in files:
            layer_vo = analyzer._detect_layer(f, root_dir)
            if not layer_vo:
                continue
            definition = analyzer.layer_map[layer_vo]

            if analyzer.parser.is_barrel_file(f):
                self._check_barrel_completeness(
                    f, definition, layer_vo, analyzer, results
                )
                continue

            self._check_forbid_internal_all(f, definition, analyzer, results)
            self._check_no_primitives(f, definition, analyzer, results)

    def _check_barrel_completeness(
        self,
        f: FilePath,
        definition,
        layer_name: LayerNameVO,
        analyzer,
        results: LintResultList,
    ) -> None:
        if definition.barrel_completeness and not analyzer.parser.has_all_export(f):
            default_msg = LintMessage(value="__init__.py missing __all__ export list.")
            violation_msg = definition.barrel_completeness_violation_message
            results.append(
                LintResult(
                    file=f,
                    line=LineNumber(value=0),
                    column=ColumnNumber(value=0),
                    code=ErrorCode(code="AES012"),
                    message=violation_msg or default_msg,
                    source=AdapterName(value="architecture"),
                    severity=Severity.MEDIUM,
                )
            )

    def _check_forbid_internal_all(
        self, f: FilePath, definition, analyzer, results: LintResultList
    ) -> None:
        if definition.forbid_internal_all and analyzer.parser.has_all_export(f):
            default_msg = LintMessage(value="__all__ is forbidden in non-barrel files.")
            violation_msg = definition.forbid_internal_all_violation_message
            results.append(
                LintResult(
                    file=f,
                    line=LineNumber(value=0),
                    column=ColumnNumber(value=0),
                    code=ErrorCode(code="AES013"),
                    message=violation_msg or default_msg,
                    source=AdapterName(value="architecture"),
                    severity=Severity.MEDIUM,
                )
            )

    def _check_no_primitives(
        self, f: FilePath, definition, analyzer, results: LintResultList
    ) -> None:
        if not definition.no_primitives:
            return

        target_primitives = PRIMITIVE_TYPE_LIST
        if isinstance(definition.no_primitives, PrimitiveTypeList):
            target_primitives = definition.no_primitives

        code = ErrorCode(code="AES006")

        violations = analyzer.parser.find_primitive_violations(f, target_primitives)
        for violation in violations.values:
            default_msg = LintMessage(
                value="Use Value Objects instead of raw primitives."
            )
            violation_msg = definition.no_primitives_violation_message
            results.append(
                LintResult(
                    file=f,
                    line=violation.line,
                    column=violation.column,
                    code=code,
                    message=violation_msg or default_msg,
                    source=AdapterName(value="architecture"),
                    severity=Severity.MEDIUM,
                )
            )

File: lint-arwaky/src-rust/capabilities/arch_internal_checker.rs
// arch_internal_checker — Internal architectural rule checks (barrels, primitives).
// Implements IInternalCheckerProtocol: barrel completeness, forbid_internal_all, no_primitives.

use std::fs;
use crate::taxonomy::{
    AdapterName, ColumnNumber, ErrorCode, FilePath, LayerDefinition,
    LayerNameVO, LintMessage, LintResult, LineNumber, Severity,
    ScopeRef, LocationList,
};

pub struct ArchInternalChecker;

impl ArchInternalChecker {
    pub fn new() -> Self {
        Self
    }

    fn make_result(file: &str, code: &str, msg: &str, sev: Severity) -> LintResult {
        LintResult {
            file: FilePath::new(file.to_string()),
            line: LineNumber::new(0),
            column: ColumnNumber::new(0),
            code: ErrorCode::new(code),
            message: LintMessage::new(msg),
            source: AdapterName::new("architecture"),
            severity: sev,
            enclosing_scope: ScopeRef {
                name: "".to_string(),
                kind: "".to_string(),
                file: FilePath::new(""),
                start_line: LineNumber::new(0),
                end_line: LineNumber::new(0),
            },
            related_locations: LocationList::new(Vec::new()),
        }
    }

    fn file_has_all_export(file: &str) -> bool {
        if let Ok(content) = fs::read_to_string(file) {
            // Python: __all__ = [...], Rust: pub use ..., JS/TS: export *
            return content.contains("__all__")
                || content.contains("pub use")
                || content.contains("export *")
                || content.contains("export {");
        }
        false
    }

    fn is_barrel_file(filename: &str) -> bool {
        matches!(filename, "__init__.py" | "mod.rs" | "index.ts" | "index.js")
    }

    /// Check barrel completeness (AES012): barrel files must have __all__ / pub use.
    pub fn check_barrel_completeness(
        &self,
        file: &str,
        filename: &str,
        definition: &LayerDefinition,
        violations: &mut Vec<LintResult>,
    ) {
        if !definition.barrel_completeness.value {
            return;
        }
        if !Self::file_has_all_export(file) {
            let msg = if !definition.barrel_completeness_violation_message.value.is_empty() {
                definition.barrel_completeness_violation_message.value.clone()
            } else {
                "__init__.py missing __all__ export list.".to_string()
            };
            violations.push(Self::make_result(file, "AES012", &msg, Severity::MEDIUM));
        }
    }

    /// Check forbid_internal_all (AES013): non-barrel files must NOT have __all__.
    pub fn check_forbid_internal_all(
        &self,
        file: &str,
        definition: &LayerDefinition,
        violations: &mut Vec<LintResult>,
    ) {
        if !definition.forbid_internal_all.value {
            return;
        }
        if Self::file_has_all_export(file) {
            let msg = if !definition.forbid_internal_all_violation_message.value.is_empty() {
                definition.forbid_internal_all_violation_message.value.clone()
            } else {
                "__all__ is forbidden in non-barrel files.".to_string()
            };
            violations.push(Self::make_result(file, "AES013", &msg, Severity::MEDIUM));
        }
    }

    /// Check internal rules for a single file (barrel completeness or forbid_internal_all + no_primitives).
    pub fn check_internal_rules(
        &self,
        file: &str,
        filename: &str,
        definition: Option<&LayerDefinition>,
        violations: &mut Vec<LintResult>,
    ) {
        let def = match definition {
            Some(d) => d,
            None => return,
        };

        if Self::is_barrel_file(filename) {
            self.check_barrel_completeness(file, filename, def, violations);
            return;
        }

        self.check_forbid_internal_all(file, def, violations);
        // Note: no_primitives check (AES006) requires AST parsing of class attributes.
        // That is delegated to the main ArchitectureRulesEvaluator which has AST access.
    }
}
